### PR TITLE
feat(render-worker): frame caching to eliminate redundant text rendering

### PR DESCRIPTION
**Pull request by mhuang74**
## Summary

Add full-frame byte caching to `FrameRenderer` to eliminate redundant CJK text rendering across consecutive frames. Lyrics only change ~200 times in a 20-minute video, but the render worker currently re-renders ALL text from scratch every frame (~28,800 times), causing Lambda timeouts for 4+ song renders.

## Performance Impact

| Metric | Before | After |
|--------|--------|-------|
| Text render ops (20-min video) | 28,800 | ~200 + ~32 fade steps |
| Per-frame cost (cached hit) | ~40ms | ~0.05ms |
| Frame rendering time (20-min video) | ~19.2 min | ~31s |
| Total pipeline (4 songs) | ~15+ min (times out!) | ~5-7 min (fits in Lambda) |

### `frame_renderer.py` — Core caching logic
- **`VisualState` dataclass**: Single source of truth for visual state resolution, consumed by both cache key computation and rendering
- **`_resolve_visual_state()`**: Extracts segment/lyric/alpha resolution from `render_frame()` into a shared method
- **`_compute_cache_key()`**: Deterministic tuple key from visual state (segment_id, title, lyric_index, quantized alpha values)
- **`_quantize_alpha()`**: Reduces unique fade frames from ~100 to ~16 per fade event (configurable via `SOW_FADE_ALPHA_STEPS`, default 16)
- **`render_frame_bytes()`**: New method returning `bytes` with LRU `OrderedDict` cache — on hit returns cached bytes directly (zero copies), on miss renders + `tobytes()` + stores
- **`_render_frame_impl()`**: Extracted rendering body shared by both `render_frame()` (returns `Image.Image`, backward compatible) and `render_frame_bytes()`
- **`_compute_intro_alpha()` / `_compute_last_lyric_fade_alpha()`**: Extracted alpha computation from `render_intro_info()` and `render_lyrics()` for reuse by both caching and rendering
- **`clear_cache()` / `get_cache_stats()`**: Cache management and observability

### `video_engine.py` — Encoding loop update
- Switch from `render_frame()` + `img.tobytes()` to `render_frame_bytes()` — eliminates `tobytes` timing tracking
- Add periodic cache stats logging (every 5s of video) with hit rate percentage
- Add final cache stats logging in `finally` block
- Remove `tobytes_total_ns` from encoding breakdown logs

### `deploy/aws-up.sh` — Lambda deployment config
- Add `SOW_FRAME_CACHE_ENABLED`, `SOW_FADE_ALPHA_STEPS`, `SOW_MAX_CACHE_ENTRIES` env vars
- Increase Lambda memory from 2048MB to 3072MB (cache needs ~2.5GB peak)

### `docker-compose.yml` — Local dev config
- Add same cache env vars with defaults
- Update memory to 3072MB

### `test_frame_renderer.py` — 98 tests (all passing)
- `TestGetBoolEnv` / `TestGetIntEnv` — env var parsing edge cases
- `TestQuantizeAlpha` — alpha quantization boundary values, custom step sizes
- `TestVisualState` — visual state resolution for intro, lyrics, fade, segment transitions
- `TestComputeCacheKey` — determinism, different states produce different keys
- `TestFrameCache` — hit/miss behavior, LRU eviction, bytes match `render_frame().tobytes()`
- `TestCacheDisabled` — no caching when `SOW_FRAME_CACHE_ENABLED=false`
- `TestFrameCachePerformance` — benchmark: cached calls significantly faster than uncached
- `TestRenderFrameBytesBackwardCompat` — `render_frame()` still returns `Image.Image`, bytes match

### `test_video_engine.py`
- Update `test_encode_lyrics_timeline_sync_with_title_card` to patch `render_frame_bytes` instead of `render_frame`

## Memory Budget (3GB Lambda)

| Component | Memory |
|-----------|--------|
| Python runtime + libraries | ~100 MB |
| FFmpeg subprocess + buffers | ~50-100 MB |
| Audio files on disk | ~50-100 MB |
| Frame cache (300 entries × 8.3MB) | ~2,490 MB |
| **Total worst case** | **~2,745 MB** |
| **Lambda limit** | **3,072 MB** |
| **Headroom** | **~327 MB** |

## Configuration (Env Vars)

| Variable | Default | Purpose |
|----------|---------|---------|
| `SOW_FRAME_CACHE_ENABLED` | `true` | Kill switch — disable caching without redeploy |
| `SOW_FADE_ALPHA_STEPS` | `16` | Fade quantization steps (higher = smoother, more memory) |
| `SOW_MAX_CACHE_ENTRIES` | `300` | LRU cache cap (each entry ≈ 8.3MB) |

## Testing

All 98 frame renderer tests and 51 video engine tests pass. The performance benchmark test confirms cached calls are significantly faster than uncached calls.

### CODE REVIEW

File: services/render-worker/src/sow_render_worker/video_engine.py [hunk 1]
```diff
@@ -261,7 +261,7 @@ def encode_video_with_ffmpeg(
     ) -> None:
         width, height = self.resolution
 
-        ffmpeg_start = time.monotonic()
+        ffmpeg_start_ns = time.monotonic_ns()
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/render-worker/src/sow_render_worker/video_engine.py:267]: Fixed in 546b5b6. Added `self.frame_renderer.clear_cache()` at the start of `encode_video_with_ffmpeg` to prevent cross-job cache contamination in warm Lambda containers.

File: services/render-worker/src/sow_render_worker/frame_renderer.py [hunk 2]
```diff
@@ -130,9 +168,19 @@ def __init__(
         self.resolution = resolution or template.resolution
         self.base_font_size = FONT_SIZE_PRESETS[font_size_preset]
 
+        self._cache_enabled = _get_bool_env("SOW_FRAME_CACHE_ENABLED", _DEFAULT_CACHE_ENABLED)
+        self._fade_alpha_steps = max(2, _get_int_env("SOW_FADE_ALPHA_STEPS", _DEFAULT_FADE_ALPHA_STEPS))
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/render-worker/src/sow_render_worker/frame_renderer.py]: Fixed in 546b5b6. Capped `_fade_alpha_steps` at 256: `min(256, max(2, ...))` to prevent `_alpha_step_size` from becoming 0 and causing ZeroDivisionError in `_quantize_alpha`.

File: services/render-worker/src/sow_render_worker/frame_renderer.py [hunk 3]
```diff
@@ -130,9 +168,19 @@ def __init__(
...
+        self._max_cache_entries = max(10, _get_int_env("SOW_MAX_CACHE_ENTRIES", _DEFAULT_MAX_CACHE_ENTRIES))
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/render-worker/src/sow_render_worker/frame_renderer.py]: Fixed in 546b5b6. Lowered minimum from `max(10, ...)` to `max(1, ...)` so `test_lru_eviction` with `SOW_MAX_CACHE_ENTRIES=2` works correctly.

File: services/render-worker/src/sow_render_worker/frame_renderer.py [hunk 4]
```diff
@@ -189,20 +309,123 @@ def render_frame(
...
+
+        if current_song_lyrics and current_time >= current_song_lyrics[0].global_time_seconds:
+            for i, line in enumerate(current_song_lyrics):
+                if line.global_time_seconds <= current_time:
+                    current_lyric_index = i
+                else:
+                    break
+
+            if current_lyric_index < 0 and current_time > current_song_lyrics[-1].global_time_seconds:
+                current_lyric_index = len(current_song_lyrics) - 1
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply services/render-worker/src/sow_render_worker/frame_renderer.py]: Fixed in 546b5b6. Removed the dead `if current_lyric_index < 0 and ...` block since the outer guard guarantees the loop always finds at least index 0.
